### PR TITLE
Shadowed Predeclared Indentifier analyzer

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Tool

**What does this PR do? Why is it needed?**

This PR introduces a static code analyzer that reports a diagnostic when a predefined Go identifier (e.g. `int`, `error`, `len`) has been shadowed by another identifier. Occurrences of such code in Prysm have been found here: https://deepsource.io/gh/rkapka/prysm/issue/CRT-A0001/occurrences.

It also fixes issues found by the analyzer.

**Which issues(s) does this PR fix?**

Part of #6747 

**Other notes for review**

The analyzer currently incorrectly reports a diagnostic if an offending name is reassigned along with another variable declaration. This is not a problem because the diagnostic will go away after the offending name declaration has been corrected. Example:
```
// Before fixing (2 diagnostics reported)
var error int  // A diagnostic will be correctly reported here
other, error := 0, 0 // Another diagnostic will be reported here, but `error` has already been covered above

// After fixing (no diagnostics reported, which is correct)
var i int
other, i := 0, 0
```

### CODE REVIEW

File: tools/analyzers/shadowpredecl/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// Doc explaining the tool.
+const Doc = "Tool to detect declarations that shadow predeclared identifiers by having the same name."
+
+const messageTemplate = "%s '%s' shadows a predeclared identifier with the same name. Choose another name."
+
+var predeclared = []string{"true", "false", "iota", "nil", "append", "cap", "close", "complex" /*"copy",*/, "delete", "imag",
```
Comment (rkapka) [inline tools/analyzers/shadowpredecl/analyzer.go:18]: I commented out `copy` because it is a common function name pattern. However only non-exported functions will conflict with the built-in identifier, so maybe let's just rename such occurrences in our codebase?

File: beacon-chain/state/types.go [hunk 2]
```diff
@@ -190,7 +190,7 @@ func newValHandler(vals []*ethpb.Validator) *validatorMapHandler {
 }
 
 // copies the whole map and returns a map handler with the copied map.
-func (v *validatorMapHandler) copy() *validatorMapHandler {
+func (v *validatorMapHandler) copyHandler() *validatorMapHandler {
```
Comment (rkapka) [inline beacon-chain/state/types.go:193]: I was hesitating whether diagnostics should be reported for `copy` because it's a common function naming pattern. I eventually decided that name conflicts will only happen for non-exported functions, for which proper naming is a little bit less important.

And `copy` is a Go built-in function after all.

File: tools/analyzers/shadowpredecl/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    importpath = "github.com/prysmaticlabs/prysm/tools/analyzers/shadowpredecl",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@org_golang_x_tools//go/analysis:go_tool_library",
+        "@org_golang_x_tools//go/analysis/passes/inspect:go_tool_library",
+        "@org_golang_x_tools//go/ast/inspector:go_tool_library",
+    ],
+)
+
+# gazelle:exclude analyzer_test.go
```
Comment (rkapka) [inline tools/analyzers/shadowpredecl/BUILD.bazel:28]: Excluded because of #7074.

File: shared/event/feed.go [hunk 4]
```diff
@@ -219,8 +219,8 @@ func (cs caseList) find(channel interface{}) int {
 	return -1
 }
 
-// delete removes the given case from cs.
-func (cs caseList) delete(index int) caseList {
+// remove removes the given case from cs.
+func (cs caseList) remove(index int) caseList {
```
Comment (farazdagi) [inline shared/event/feed.go:223]: Again, I am not sure why the change?

**Review by farazdagi — CHANGES_REQUESTED** [review]

File: beacon-chain/state/types.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/state/types.go:193]: +1 , i dont really see any benefit to changing common method names/patterns here. This just adds in an extra layer of verbosity when trying to understand the method.

Comment (rkapka) [reply beacon-chain/state/types.go:193]: This will be reverted. I did not take into consideration that functions with receiver should be treated differently because you cannot access such actions directly. And therefore they do not shadow anything.

**Review by farazdagi — APPROVED** [review]
Looks great, thank you!
